core/controllers/classifier.py

```diff
@@ -0,0 +1,125 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Controllers for communicating with the VM for training classifiers."""
+
+import hashlib
+import hmac
+import json
+
+from core.controllers import base
+from core.domain import classifier_services
+from core.domain import config_domain
+
+import feconf
+
+
+# NOTE TO DEVELOPERS: This function should be in sync with its counterpart on
+# Oppia-ml.
+def generate_signature(secret, message):
+    """Generates digital signature for given data.
+
+    Args:
+        secret: str. The secret used to communicate with Oppia-ml.
+        message: dict. The message payload data.
+
+    Returns:
+        str. The signature of the payload data.
+    """
+    message_json = json.dumps(message, sort_keys=True)
+    return hmac.new(secret, message_json, digestmod=hashlib.sha256).hexdigest()
+
+
+def validate_job_result_message_dict(message):
+    """Validates the data-type of the message payload data.
+
+    Args:
+        message: dict. The message payload data.
+
+    Returns:
+        bool. Whether the payload dict is valid.
+    """
+    job_id = message.get('job_id')
+    classifier_data = message.get('classifier_data')
+
+    if not isinstance(job_id, basestring):
+        return False
+    if not isinstance(classifier_data, dict):
+        return False
+    return True
+
+
+def verify_signature(message, vm_id, received_signature):
+    """Function that checks if the signature received from the VM is valid.
+
+    Args:
+        message: dict. The message payload data.
+        vm_id: str. The ID of the VM instance.
+        received_signature: str. The signature received from the VM.
+
+    Returns:
+        bool. Whether the incoming request is valid.
+    """
+    secret = None
+    for val in config_domain.VMID_SHARED_SECRET_KEY_MAPPING.value:
+        if val['vm_id'] == vm_id:
+            secret = str(val['shared_secret_key'])
+            break
+    if secret is None:
+        return False
+
+    generated_signature = generate_signature(secret, message)
+    if generated_signature != received_signature:
+        return False
+    return True
+
+
+class TrainedClassifierHandler(base.BaseHandler):
+    """This handler stores the result of the training job in datastore and
+    updates the status of the job.
+    """
+
+    REQUIRE_PAYLOAD_CSRF_CHECK = False
+
+    def post(self):
+        """Handles POST requests."""
+        signature = self.payload.get('signature')
+        message = self.payload.get('message')
+        vm_id = self.payload.get('vm_id')
+        if vm_id == feconf.DEFAULT_VM_ID and not feconf.DEV_MODE:
+            raise self.UnauthorizedUserException
+
+        if not validate_job_result_message_dict(message):
+            raise self.InvalidInputException
+        if not verify_signature(message, vm_id, signature):
+            raise self.UnauthorizedUserException
+
+        job_id = message.get('job_id')
```

Comment:
Let's do message['job_id'] here, and similarly for the next line. You've already validated the message so there's no chance of a KeyError, right? A .get() seems to leave open the possibility of None and I think it's best to be strict where possible.